tox.ini

```diff
@@ -1,61 +1,137 @@
-basepython = python2.7
-# Reactive source charms are Python3-only, but a py27 unit test target
-# is required by OpenStack Governance.  Remove this shim as soon as
-# permitted.  https://governance.openstack.org/tc/reference/cti/python_cti.html
-whitelist_externals = true
-commands = true
-
-[testenv:py34]
-basepython = python3.4
+commands = stestr run --slowest {posargs}
+whitelist_externals = juju
+passenv = HOME TERM CS_* OS_* TEST_*
 deps = -r{toxinidir}/test-requirements.txt
-commands = stestr run {posargs}
 
 [testenv:py35]
 basepython = python3.5
-deps = -r{toxinidir}/test-requirements.txt
-commands = stestr run {posargs}
+deps = -r{toxinidir}/requirements.txt
+       -r{toxinidir}/test-requirements.txt
 
 [testenv:py36]
 basepython = python3.6
-deps = -r{toxinidir}/test-requirements.txt
-commands = stestr run {posargs}
+deps = -r{toxinidir}/requirements.txt
+       -r{toxinidir}/test-requirements.txt
+
+[testenv:py37]
+basepython = python3.7
+deps = -r{toxinidir}/requirements.txt
+       -r{toxinidir}/test-requirements.txt
+
+[testenv:py38]
+basepython = python3.8
+deps = -r{toxinidir}/requirements.txt
+       -r{toxinidir}/test-requirements.txt
+
+[testenv:py3]
+basepython = python3
+deps = -r{toxinidir}/requirements.txt
+       -r{toxinidir}/test-requirements.txt
 
 [testenv:pep8]
 basepython = python3
-deps = -r{toxinidir}/test-requirements.txt
-commands = flake8 {posargs} src unit_tests
+deps = flake8==3.9.2
+       charm-tools==2.8.3
+commands = flake8 {posargs} unit_tests src
+           charm-proof
+
+[testenv:cover]
+# Technique based heavily upon
+# https://github.com/openstack/nova/blob/master/tox.ini
+basepython = python3
+deps = -r{toxinidir}/requirements.txt
+       -r{toxinidir}/test-requirements.txt
+setenv =
+    {[testenv]setenv}
+    PYTHON=coverage run
+commands =
+    coverage erase
+    stestr run --slowest {posargs}
+    coverage combine
+    coverage html -d cover
+    coverage xml -o cover/coverage.xml
+    coverage report
+
+[coverage:run]
+branch = True
+concurrency = multiprocessing
+parallel = True
+source =
+    .
+omit =
+    .tox/*
+    */charmhelpers/*
+    unit_tests/*
 
 [testenv:venv]
 basepython = python3
 commands = {posargs}
 
+[testenv:func-noop]
+basepython = python3
+commands =
+    functest-run-suite --help
+
+[testenv:func]
+basepython = python3
+commands =
+    functest-run-suite --keep-model
+
+[testenv:func-smoke]
+basepython = python3
+commands =
+    functest-run-suite --keep-model --smoke
+
+[testenv:func-dev]
+basepython = python3
+commands =
+    functest-run-suite --keep-model --dev
+
+[testenv:func-target]
+basepython = python3
+commands =
+    functest-run-suite --keep-model --bundle {posargs}
+
 [flake8]
-# E402 ignore necessary for path append before sys module import in actions
-extend-ignore = E402
+ignore = E402,E226,W503,W504
+exclude = */charmhelpers
+
+[testenv:build]
+basepython = python3
+commands =
+    charm build --log-level DEBUG -o {toxinidir}/build src {posargs}
```

Comment:
I get this warning when running this target:

```
$ tox -e build
build create: /home/freyes/Projects/charms/openstack/cinder-lvm/.tox/build
build installdeps: -r/home/freyes/Projects/charms/openstack/cinder-lvm/test-requirements.txt
build installed: attrs==21.2.0,autopage==0.4.0,charmhelpers==0.20.22,charms.openstack @ git+https://github.com/openstack/charms.openstack.git@6c8eb01162edef5e3f8e3af6cb90905e8b2a3065,charms.reactive==1.4.1,cliff==3.9.0,cmd2==2.1.2,colorama==0.4.4,coverage==5.5,extras==1.0.0,fixtures==3.0.0,flake8==3.9.2,future==0.18.2,Jinja2==3.0.1,MarkupSafe==2.0.1,mccabe==0.6.1,mock==4.0.3,netaddr==0.8.0,netifaces==0.11.0,os-testr==2.0.0,pbr==5.6.0,prettytable==2.2.0,psutil==5.8.0,pyaml==21.8.3,pycodestyle==2.7.0,pyflakes==2.3.1,pyparsing==2.4.7,pyperclip==1.8.2,python-subunit==1.4.0,PyYAML==5.4.1,six==1.16.0,stestr==3.2.0,stevedore==3.4.0,Tempita==0.5.2,testtools==2.5.0,voluptuous==0.12.1,wcwidth==0.2.5
build run-test-pre: PYTHONHASHSEED='0'
build run-test: commands[0] | charm build --log-level DEBUG -o /home/freyes/Projects/charms/openstack/cinder-lvm/build src
WARNING: test command found but not installed in testenv
  cmd: /snap/bin/charm
  env: /home/freyes/Projects/charms/openstack/cinder-lvm/.tox/build
Maybe you forgot to specify a dependency? See also the allowlist_externals envconfig setting.

DEPRECATION WARNING: this will be an error in tox 4 and above!
```

And the build charm gets the wrong directory name:

```
$ ls build/
src
$ ls build/src/
actions       bin          copyright                exec.d  layer.yaml  LICENSE   metadata.yaml  README.md         templates              tests    version     wheelhouse.txt
actions.yaml  config.yaml  copyright.layer-options  hooks   lib         Makefile  reactive       requirements.txt  test-requirements.txt  tox.ini  wheelhouse
```

Reply:
Made a blunder when updating some config parameters. I've fixed it now, should be all green.